### PR TITLE
Added jira endpoint

### PR DESCRIPTION
**Pull request by samhotep**
## Done

 - Created an endpoint to create issues for sites page requests, and save these requests to the database

### Setup
- Create a JIRA token for your account at https://id.atlassian.com/manage-profile/security/api-tokens 
- Select or create a new epic to be the parent of new issues, and copy its key e.g WD-12643
- Update the .env file with the JIRA details 
- Create a user in the db or edit one to use an email which exists in JIRA e.g samuel@canonical.com

### Requests
Make requests to `/request-changes` with a similar payload to the below
```python
            {
                "due_date": "2022-01-01",
                "reporter_id": 1, # user id
                "webpage_id": 31,
                "type": 1,
                "description": "This is a description",
            }
```
- Make a new webpage request (type: 0)
  - A subtask (issue) should be created for copy updates on the new epic
- Make a page refresh request to update an existing webpage (type: 1)
  - A **new** epic should be created, with three subtasks for UX, Visual and Dev
- Make a request to create a new webpage (type: 2)
  - A **new** epic should be created, with three subtasks for UX, Visual and Dev

## Fixes

 - Fixes [WD-13359](https://warthogs.atlassian.net/browse/WD-13359?atlOrigin=eyJpIjoiM2JjMzA2NDBhODdlNGJhMGEzMDg0NDM4ZmRhMjAyZjciLCJwIjoiaiJ9)

[WD-13359]: https://warthogs.atlassian.net/browse/WD-13359?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webapp/jira.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                raise Exception
+        except Exception:
+            raise Exception(f"Webpage with ID {webpage_id} not found")
+
+        # Get the reporter ID
+        reporter = self.get_reporter_jira_id(reporter)
+
+        # Determine the correct issue type
+        if issue_type == 0:
+            parent = None
```
Comment (akbarkz) [inline webapp/jira.py:225]: if it's a copy update type, a parent epic's key should be obtained from the env variable

Comment (akbarkz) [reply webapp/jira.py:225]: the logic you implement in this block should actually be applied to types 1 and 2. So for copy updates you create a task and attach to an existing parent, while as for page refresh and new webpage types you create an epic and 3 tasks inside that epic.

**Review by akbarkz — CHANGES_REQUESTED** [review]

File: webapp/jira.py (hunk 1 above)
Comment (samhotep) [reply webapp/jira.py:225]: I see, let me in fact make separate globals for the issue types that indicate which one is which

File: webapp/jira.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        # Get the webpage
+        try:
+            webpage = Webpage.query.filter_by(id=webpage_id).first()
+            if not webpage:
+                raise Exception
+        except Exception:
+            raise Exception(f"Webpage with ID {webpage_id} not found")
+
+        # Get the reporter ID
+        reporter = self.get_reporter_jira_id(reporter)
```
[the remark this answers is not in the document]
Comment (samhotep) [reply webapp/jira.py:221]: good catch. I'll update it

**Comment by akbarkz** [comment]
I received an error with the following text: `User with email [akbar.abdrakhmanov@canonical.com](mailto:akbar.abdrakhmanov@canonical.com) not found in Jira` that doesn't sound right, since that's my email in Jira.

File: webapp/helper.py [hunk 3]
```diff
@@ -1,18 +1,45 @@
...
+        webpage_id=task["webpage_id"],
+        request_type=task["type"],
+        description=task["description"],
+    )
+
+    # Create jira task in the database
+    get_or_create(
+        db.session,
+        JiraTask,
+        jira_id=issue["id"],
```
Comment (akbarkz) [inline webapp/helper.py:42]: I'm afraid, this identifier is not useful for us, we need to store a key of a task, like WD-12345

File: webapp/models.py [hunk 4]
```diff
@@ -121,7 +129,7 @@ class JiraTask(db.Model, DateTimeMixin):
     jira_id: int = Column(Integer)
     webpage_id: int = Column(Integer, ForeignKey("webpages.id"))
     user_id: int = Column(Integer, ForeignKey("users.id"))
-    status: str = Column(String)  # Will be filled from Jira API
+    status: str = Column(Enum(JIRATaskStatus), default=JIRATaskStatus.TRIAGED)
     created_at: str = Column(String)
```
Comment (akbarkz) [inline webapp/models.py:133]: this column stays `null` after a new row is added

File: webapp/models.py [hunk 5]
```diff
@@ -121,7 +129,7 @@ class JiraTask(db.Model, DateTimeMixin):
     jira_id: int = Column(Integer)
```
Comment (akbarkz) [inline webapp/models.py:129]: since this should be a key, the type should be String

File: webapp/helper.py (hunk 3 above)
Comment (samhotep) [reply webapp/helper.py:42]: I've made the update to use the key

File: webapp/models.py (hunk 4 above)
Comment (samhotep) [reply webapp/models.py:133]: I've removed the enum, and now using plain strings

File: webapp/models.py (hunk 5 above)
Comment (samhotep) [reply webapp/models.py]: Done 👍🏿

File: webapp/models.py (hunk 4 above)
Comment (akbarkz) [reply webapp/models.py:133]: @samhotep this is still an issue, after a row is added to "jira_tasks" table, the "created_at_ column is `null`

File: webapp/jira.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+                "summary": summary,
+                "issuetype": {"id": issue_type},
+                "labels": self.labels,
+                "reporter": {"id": reporter_jira_id},
+                "duedate": due_date,
+                "parent": parent,
+                "project": {"id": "10492"},  # Web and Design-ENG
+                "components": [
+                    {"id": "12655"},  # Sites Tribe
+                    {"id": "12628"},  # Infrastructure
```
Comment (akbarkz) [inline webapp/jira.py:162]: please remove "Infrastructure"

Comment (samhotep) [reply webapp/jira.py:162]: Alright.

File: webapp/models.py (hunk 4 above)
Comment (samhotep) [reply webapp/models.py]: I've fixed it, removing the extra created_at declaration did the trick

**Review by akbarkz — APPROVED** [review]

**Pull request merged**